xbmc/windowing/gbm/drm/DRMCrtc.h

```diff
@@ -0,0 +1,45 @@
+/*
+ *  Copyright (C) 2005-2020 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "DRMObject.h"
+
+namespace KODI
+{
+namespace WINDOWING
+{
+namespace GBM
+{
+
+class CDRMCrtc : public CDRMObject
+{
+public:
+  explicit CDRMCrtc(int fd, uint32_t crtc);
+  CDRMCrtc(const CDRMCrtc&) = delete;
+  CDRMCrtc& operator=(const CDRMCrtc&) = delete;
+  ~CDRMCrtc() = default;
+
+  uint32_t& GetCrtcId() const { return m_crtc->crtc_id; }
```

Comment:
It's a bit misleading to have a const function return a mutable result, as the function can be used in a const context, but the object can be mutated without const checks by assigning to the result reference, bypassing const-correctness.

For better const-correctness, the `const` here should be dropped. If const access is needed, then the idiomatic approach is to have two functions, one for each constness (one const and one mutable)

Reply:
I never thought about it this way.
For my understanding the const just makes sure nobody accidently modifies the class state in such method.

Reply:
How about this:

```c++
uint32_t GetCrtcId() const { return m_crtc->crtc_id; }
void SetCrtcId(uint32_t crtcId) { m_crtc->crtc_id = crtcId; }
```

Reply:
The setter is not needed. Basically I'm providing getters for the underlying C-struct that shouldn't be modified.

Reply:
You shouldn't pass uint32_t by reference. the compiler can perform drastically superior optimizations of ints. I think what you want is:

```c++
uint32_t GetCrtcId() const { return m_crtc->crtc_id; }
```

Reply:
Yeah I see 2 possibilities:
`uint32_t GetCrtcId() const { return m_crtc->crtc_id; } // doesn't prevent developer from changing his crtc_id, but never changes the value in the class instance (you hand out a copy)`

Taken from your linked article:
`std::remove_reference_t<uint32_t> const uint32_t& GetCrtcId() const { return m_crtc->crtc_id; }`

Should give you an immutable reference, haven't tried though.

Reply:
imho just remove the ref and use plain int instead.